File: lib/core/category_resolver.dart
/// Pure category-resolution helpers shared by the ingest paths (carrier sync in
/// the app, and — mirrored by hand — server/src/lib/category_resolver.ts and
/// supabase/functions/_shared/category_resolver.ts). Kept free of
/// Supabase/Flutter so the priority rules are unit-testable.
///
/// Store and item categories are resolved independently:
///   * An item's category comes only from item-level signals — its own history,
///     then the keyword rules run against the item name. It never inherits the
///     store/merchant category (that signal is unreliable at the item level).
///   * The store (invoice header) category comes from the merchant's history,
///     then the keyword rules run against the merchant name, and finally — when
///     neither matches — the most common category among its line items.
/// History always consults the user's own past choices first, and the *most
/// recent* past choice wins.
library;

String _norm(String s) => s.trim().toLowerCase();

/// One history row: a lookup [key] (item name or merchant), the [categoryId] the
/// user assigned (null → no signal, skipped), and an ISO recency [stamp] where a
/// lexicographically-larger string is more recent (e.g. 'invoice_date|created_at').
typedef HistoryRow = ({String? key, int? categoryId, String stamp});

/// Reduces history [rows] to one category id per normalized key, keeping the most
/// recent. Rows with a null key/category or blank key are ignored.
Map<String, int> foldMostRecentCategory(Iterable<HistoryRow> rows) {
  final best = <String, ({String stamp, int catId})>{};
  for (final r in rows) {
    final catId = r.categoryId;
    final rawKey = r.key;
    if (catId == null || rawKey == null) continue;
    final key = _norm(rawKey);
    if (key.isEmpty) continue;
    final cur = best[key];
    if (cur == null || r.stamp.compareTo(cur.stamp) > 0) {
      best[key] = (stamp: r.stamp, catId: catId);
    }
  }
  return {for (final e in best.entries) e.key: e.value.catId};
}

/// Category for a line item: its own history → [keywordFallback] (the keyword
/// rules run against this item's name). Never falls back to the store/merchant
/// category — that signal is unreliable at the item level.
int? resolveItemCategory({
  required String itemName,
  required Map<String, int> itemHistory,
  required int? keywordFallback,
}) =>
    itemHistory[_norm(itemName)] ?? keywordFallback;

/// Category for an invoice header (the whole receipt): merchant history →
/// [keywordFallback] (the keyword rules run against the merchant name) → the
/// most common category among [itemCategoryIds] (its already-resolved line
/// items). The item-mode fallback covers stores the rules don't recognise by
/// name but whose items are categorizable.
int? resolveInvoiceCategory({
  required String? merchant,
  required Map<String, int> merchantHistory,
  required int? keywordFallback,
  required Iterable<int?> itemCategoryIds,
}) =>
    merchantHistory[_norm(merchant ?? '')] ??
    keywordFallback ??
    modeCategory(itemCategoryIds);

/// The most frequent non-null id in [ids], or null when there are none. Ties are
/// broken by first appearance: [counts] iterates in first-insertion order (= the
/// order each id first appears in [ids]), so a strict `>` that never replaces on
/// a tie keeps the earliest-seen id — deterministic when [ids] is in `sort_order`.
int? modeCategory(Iterable<int?> ids) {
  final counts = <int, int>{};
  for (final id in ids) {
    if (id != null) counts[id] = (counts[id] ?? 0) + 1;
  }
  int? best;
  var bestCount = 0;
  for (final entry in counts.entries) {
    if (entry.value > bestCount) {
      best = entry.key;
      bestCount = entry.value;
    }
  }
  return best;
}

File: supabase/functions/_shared/category_resolver.ts
// Pure category-resolution helpers — the Deno mirror of
// lib/core/category_resolver.dart and server/src/lib/category_resolver.ts. Store
// and item categories are resolved independently: an item's category comes only
// from item-level signals (its own history, then the keyword rules on the item
// name) and never inherits the store/merchant category; the store (invoice
// header) category comes from the merchant's history, then the keyword rules on
// the merchant name, and finally the most common category among its line items.
// History consults the user's own past choices first; the most recent past
// choice wins.

export const norm = (s: string) => s.trim().toLowerCase();

/** One history row: a lookup key, the assigned category id (null → skipped), and
 * an ISO recency stamp where a lexicographically-larger string is more recent. */
export interface HistoryRow {
  key: string | null;
  categoryId: number | null;
  stamp: string;
}

/** Reduces history rows to one category id per normalized key, keeping the most
 * recent. Rows with a null key/category or blank key are ignored. */
export function foldMostRecentCategory(
  rows: Iterable<HistoryRow>,
): Map<string, number> {
  const best = new Map<string, { stamp: string; catId: number }>();
  for (const r of rows) {
    if (r.categoryId == null || r.key == null) continue;
    const key = norm(r.key);
    if (!key) continue;
    const cur = best.get(key);
    if (!cur || r.stamp > cur.stamp) best.set(key, { stamp: r.stamp, catId: r.categoryId });
  }
  const out = new Map<string, number>();
  for (const [k, v] of best) out.set(k, v.catId);
  return out;
}

/** Category for a line item: its own history → keyword fallback (the keyword
 * rules run against this item's name). Never falls back to the store/merchant
 * category — that signal is unreliable at the item level. */
export function resolveItemCategory(
  itemName: string,
  itemHistory: Map<string, number>,
  keywordFallback: number | null,
): number | null {
  return itemHistory.get(norm(itemName)) ?? keywordFallback;
}

/** Category for an invoice header (the whole receipt): merchant history →
 * keyword fallback (the keyword rules run against the merchant name) → the most
 * common category among its already-resolved line items. */
export function resolveInvoiceCategory(
  merchant: string | null,
  merchantHistory: Map<string, number>,
  keywordFallback: number | null,
  itemCategoryIds: Iterable<number | null>,
): number | null {
  return (
    merchantHistory.get(norm(merchant ?? "")) ??
    keywordFallback ??
    modeCategory(itemCategoryIds)
  );
}

/** The most frequent non-null id, or null when there are none. Ties are broken
 * by first appearance: `counts` iterates in first-insertion order (= the order
 * each id first appears in `ids`), so a strict `>` that never replaces on a tie
 * keeps the earliest-seen id — deterministic when `ids` is in sort_order. */
export function modeCategory(ids: Iterable<number | null>): number | null {
  const counts = new Map<number, number>();
  for (const id of ids) {
    if (id != null) counts.set(id, (counts.get(id) ?? 0) + 1);
  }
  let best: number | null = null;
  let bestCount = 0;
  for (const [id, count] of counts) {
    if (count > bestCount) {
      best = id;
      bestCount = count;
    }
  }
  return best;
}

export function distinctTrimmed(values: string[]): string[] {
  const out = new Set<string>();
  for (const v of values) {
    const t = v.trim();
    if (t) out.add(t);
  }
  return [...out];
}
